File: app/services/query_grammar/rhyme.py
"""韻／聲錨 grammar（#3 第三輪）。"""
from __future__ import annotations

import re
from typing import Optional

from app.services.query_grammar.equals import (
    is_framed_equals_query,
    is_hybrid_tail_equals_alias,
)
from app.services.query_tokens import CODE_TAIL_MIDDLE, SLOT_CHARS_RE, WILDCARD_CHARS

_RHYME_ANCHOR_SHAPE_RE = re.compile(
    rf"^(?:"
    rf"({SLOT_CHARS_RE}+)([一-龥])=$|"
    rf"([一-龥])=({SLOT_CHARS_RE}+)$|"
    rf"=([一-龥])({SLOT_CHARS_RE}+)$|"
    rf"({SLOT_CHARS_RE}+)=([一-龥])$"
    rf")"
)


def _is_wildcard_char(ch: str) -> bool:
    return len(ch) == 1 and ch in WILDCARD_CHARS


def normalize_partial_rhyme_mask_query(q: str) -> str:
    """`窮困潦=?` → `窮困潦?=`（尾格通配 + 韻錨）。"""
    m = re.fullmatch(r"^([一-龥]{3})=\?$", q)
    if m:
        return f"{m.group(1)}?="
    return q


def parse_partial_rhyme_mask_query(q: str) -> Optional[dict]:
    """四字部分韻錨：`窮?潦倒=` 等；`?` 通配，其餘格同參考字韻母。"""
    nq = normalize_partial_rhyme_mask_query(q)
    m = re.fullmatch(r"^([一-龥?]{4})=$", nq)
    if not m:
        return None
    pattern = m.group(1)
    if "?" not in pattern or all(ch == "?" for ch in pattern):
        return None
    if pattern.startswith("?") and re.fullmatch(r"\?[一-龥]{3}", pattern):
        return None
    anchors = [(pos, ch) for pos, ch in enumerate(pattern) if ch != "?"]
    if not anchors:
        return None
    return {"raw_q": q, "pattern": pattern, "width": 4, "anchors": anchors}


def blocks_star_normalize(q: str) -> bool:
    """韻／聲錨形狀 — star normalize 唔插 `*`。"""
    if parse_partial_rhyme_mask_query(q):
        return True
    return bool(_RHYME_ANCHOR_SHAPE_RE.match(q))


def parse_code_ref_rhyme_contradiction_hint(q: str) -> Optional[str]:
    """?3人? 無 = → hint（P3）。"""
    m = re.match(r"^([?_%]+)(\d+)([一-龥])([?_%])$", q)
    if m and "=" not in q:
        d, ref = m.group(2), m.group(3)
        return f"碼位同參考字「{ref}」衝突：請改用 `?{d}{ref}=?` 標中格同韻。"
    return None


def parse_code_ref_middle_rhyme_query(q: str) -> Optional[dict]:
    """碼＋參考字＋通配（中格韻）：?3人=?（P3）。"""
    m = re.match(r"^([?_%]+)(\d+)([一-龥])=\?$", q)
    if not m:
        return None
    leading, digits, anchor = m.group(1), m.group(2), m.group(3)
    width = len(leading) + len(digits) + 1
    anchor_pos = len(leading) + len(digits) - 1
    slots: list[dict] = []
    for i, d in enumerate(digits):
        slots.append({"pos": len(leading) + i, "kind": "code_digit", "value": d})
    slots.append({"pos": anchor_pos, "kind": "final_anchor", "value": anchor})
    return {
        "raw_q": q,
        "width": width,
        "anchor": anchor,
        "anchor_pos": anchor_pos,
        "leading": leading,
        "digits": digits,
        "slots": slots,
    }


def parse_double_wildcard_rhyme_query(q: str) -> Optional[dict]:
    """二字韻錨 ?*就=（P2）。"""
    m = re.match(r"^([?_%])\*([一-龥])=$", q)
    if not m:
        return None
    return {
        "constraint": "final",
        "anchor": m.group(2),
        "anchor_pos": 1,
        "slots": m.group(1),
        "width": 2,
    }


def parse_double_wildcard_initial_query(q: str) -> Optional[dict]:
    """二字聲錨 ?*=就（P2 對稱）。"""
    m = re.match(r"^([?_%])\*=([一-龥])$", q)
    if not m:
        return None
    return {
        "constraint": "initial",
        "anchor": m.group(2),
        "anchor_pos": 1,
        "slots": m.group(1),
        "width": 2,
    }


def parse_rhyme_anchor_query(q: str) -> Optional[dict]:
    """Query-level rhyme/initial anchor: 就= / =就 / 香=? / ?*就= / =香? / ?*=就."""
    if not q or CODE_TAIL_MIDDLE in q or "@" in q or is_framed_equals_query(q):
        return None
    if parse_double_wildcard_rhyme_query(q) or parse_double_wildcard_initial_query(q):
        return None

    m = re.match(r"^([一-龥])=$", q)
    if m:
        return {
            "constraint": "final",
            "anchor": m.group(1),
            "anchor_pos": 0,
            "slots": "",
            "width": 1,
        }

    m = re.match(r"^=([一-龥])$", q)
    if m:
        return {
            "constraint": "initial",
            "anchor": m.group(1),
            "anchor_pos": 0,
            "slots": "",
            "width": 1,
        }

    m = re.match(rf"^({SLOT_CHARS_RE}+)([一-龥])=$", q)
    if m:
        slots, anchor = m.group(1), m.group(2)
        width = len(slots) + 1
        return {
            "constraint": "final",
            "anchor_pos": width - 1,
            "anchor": anchor,
            "slots": slots,
            "width": width,
        }

    m = re.match(rf"^([一-龥])=({SLOT_CHARS_RE}+)$", q)
    if m:
        anchor, slots = m.group(1), m.group(2)
        width = len(slots) + 1
        return {
            "constraint": "final",
            "anchor_pos": 0,
            "anchor": anchor,
            "slots": slots,
            "width": width,
        }

    m = re.match(rf"^=([一-龥])({SLOT_CHARS_RE}+)$", q)
    if m:
        anchor, slots = m.group(1), m.group(2)
        width = len(slots) + 1
        return {
            "constraint": "initial",
            "anchor_pos": 0,
            "anchor": anchor,
            "slots": slots,
            "width": width,
        }

    m = re.match(rf"^({SLOT_CHARS_RE}+)=([一-龥])$", q)
    if m:
        slots, anchor = m.group(1), m.group(2)
        width = len(slots) + 1
        return {
            "constraint": "initial",
            "anchor_pos": width - 1,
            "anchor": anchor,
            "slots": slots,
            "width": width,
        }
    return None


def parse_triple_rhyme_anchor_query(q: str) -> Optional[dict]:
    """中格同韻三字：規範形 ?*{參考字}=?（?{字}=? normalize 後）。"""
    if not q or "@" in q or is_framed_equals_query(q):
        return None
    if is_hybrid_tail_equals_alias(q):
        return None

    m = re.match(r"^(\?\*)([一-龥])=\?$", q)
    if m:
        anchor = m.group(2)
        return {
            "anchor": anchor,
            "anchor_pos": 1,
            "width": 3,
            "leading_slots": m.group(1),
            "constraint": "final",
        }

    if CODE_TAIL_MIDDLE in q:
        return None

    m = re.match(rf"^({SLOT_CHARS_RE}+)([一-龥])=(\?)$", q)
    if not m:
        return None
    leading, anchor, _trail = m.group(1), m.group(2), m.group(3)
    if not any(_is_wildcard_char(c) for c in leading):
        return None
    if re.search(r"\d", leading):
        return None
    anchor_pos = len(leading)
    width = anchor_pos + 2
    return {
        "anchor": anchor,
        "anchor_pos": anchor_pos,
        "width": width,
        "leading_slots": leading,
        "constraint": "final",
    }

File: client/src/db/query/grammar/rhyme.ts
/** Port of query_grammar/rhyme (partial + anchors + double wild + code-ref). */
import { isWildcardChar } from '../../position-match/mask-grammar.ts';
import { QueryKind } from '../../query-kind.ts';
import type {
  CodeRefMiddleRhymeQuery,
  PartialInitialMaskQuery,
  PartialRhymeMaskQuery,
  RhymeAnchorQuery,
  TripleRhymeAnchorQuery,
} from '../../query-types.ts';
import { CODE_TAIL_MIDDLE } from './shared.ts';
import { isFramedEqualsQuery } from './equals.ts';

/** Port of rhyme.parse_code_ref_rhyme_contradiction_hint */
export function parseCodeRefRhymeContradictionHint(q: string): string | null {
  const m = q.match(/^([?_%]+)(\d+)([\u4e00-\u9fff])([?_%])$/);
  if (m && !q.includes('=')) {
    return `碼位同參考字「${m[3]}」衝突：請改用 \`?${m[2]}${m[3]}=?\` 標中格同韻。`;
  }
  return null;
}

/** Port of rhyme.parse_code_ref_middle_rhyme_query */
export function parseCodeRefMiddleRhymeQuery(q: string): CodeRefMiddleRhymeQuery | null {
  const m = q.match(/^([?_%]+)(\d+)([\u4e00-\u9fff])=\?$/);
  if (!m) {
    return null;
  }
  const leading = m[1]!;
  const digits = m[2]!;
  const anchor = m[3]!;
  const width = leading.length + digits.length + 1;
  const anchorPos = leading.length + digits.length - 1;
  const slots: CodeRefMiddleRhymeQuery['slots'] = [];
  for (let i = 0; i < digits.length; i++) {
    slots.push({ pos: leading.length + i, kind: 'code_digit', value: digits[i] });
  }
  slots.push({ pos: anchorPos, kind: 'final_anchor', value: anchor });
  return {
    kind: QueryKind.CODE_REF_MIDDLE_RHYME,
    raw_q: q,
    width,
    anchor,
    anchor_pos: anchorPos,
    leading,
    digits,
    slots,
  };
}

/** Port of rhyme.parse_double_wildcard_rhyme_query */
export function parseDoubleWildcardRhymeQuery(q: string): RhymeAnchorQuery | null {
  const m = q.match(/^([?_%])\+([\u4e00-\u9fff])=$/);
  if (!m) {
    return null;
  }
  return {
    kind: QueryKind.RHYME_ANCHOR,
    raw_q: q,
    constraint: 'final',
    anchor: m[2]!,
    anchor_pos: 1,
    slots: m[1]!,
    width: 2,
  };
}

/** Port of rhyme.parse_double_wildcard_initial_query */
export function parseDoubleWildcardInitialQuery(q: string): RhymeAnchorQuery | null {
  const m = q.match(/^([?_%])\+=([\u4e00-\u9fff])$/);
  if (!m) {
    return null;
  }
  return {
    kind: QueryKind.RHYME_ANCHOR,
    raw_q: q,
    constraint: 'initial',
    anchor: m[2]!,
    anchor_pos: 1,
    slots: m[1]!,
    width: 2,
  };
}

/** Port of rhyme.parse_triple_rhyme_anchor_query */
export function parseTripleRhymeAnchorQuery(q: string): TripleRhymeAnchorQuery | null {
  if (!q || q.includes('@') || isFramedEqualsQuery(q)) {
    return null;
  }

  let m = q.match(/^(\?\+)([\u4e00-\u9fff])=\?$/);
  if (m) {
    return {
      kind: QueryKind.TRIPLE_RHYME_ANCHOR,
      raw_q: q,
      anchor: m[2]!,
      anchor_pos: 1,
      width: 3,
      leading_slots: m[1]!,
      constraint: 'final',
    };
  }

  if (q.includes('+') || q.includes(CODE_TAIL_MIDDLE)) {
    return null;
  }

  m = q.match(/^([0-9_?%]+)([\u4e00-\u9fff])=\?$/);
  if (!m) {
    return null;
  }
  const leading = m[1]!;
  const anchor = m[2]!;
  if (![...leading].some((c) => isWildcardChar(c))) {
    return null;
  }
  if (/\d/.test(leading)) {
    return null;
  }
  const anchorPos = leading.length;
  return {
    kind: QueryKind.TRIPLE_RHYME_ANCHOR,
    raw_q: q,
    anchor,
    anchor_pos: anchorPos,
    width: anchorPos + 2,
    leading_slots: leading,
    constraint: 'final',
  };
}

/** Port of rhyme.normalize_partial_rhyme_mask_query */
function normalizePartialRhymeMaskQuery(q: string): string {
  const m = q.match(/^([\u4e00-\u9fff]{3})=\?$/);
  if (m) {
    return `${m[1]}?=`;
  }
  return q;
}

/** Port of rhyme.parse_partial_rhyme_mask_query */
export function parsePartialRhymeMaskQuery(q: string): PartialRhymeMaskQuery | null {
  const nq = normalizePartialRhymeMaskQuery(q);
  const m = nq.match(/^([\u4e00-\u9fff?]{4})=$/);
  if (!m) {
    return null;
  }
  const pattern = m[1]!;
  if (!pattern.includes('?') || pattern.split('').every((ch) => ch === '?')) {
    return null;
  }
  if (pattern.startsWith('?') && /^\?[\u4e00-\u9fff]{3}$/.test(pattern)) {
    return null;
  }
  const anchors: Array<[number, string]> = [];
  for (let pos = 0; pos < pattern.length; pos++) {
    const ch = pattern[pos]!;
    if (ch !== '?') {
      anchors.push([pos, ch]);
    }
  }
  if (!anchors.length) {
    return null;
  }
  return {
    kind: QueryKind.PARTIAL_RHYME_MASK,
    raw_q: q,
    pattern,
    width: 4,
    anchors,
  };
}

/** Port of rhyme.parse_partial_initial_mask_query */
export function parsePartialInitialMaskQuery(q: string): PartialInitialMaskQuery | null {
  const m = q.match(/^=([\u4e00-\u9fff?]{4})$/);
  if (!m) {
    return null;
  }
  const pattern = m[1]!;
  if (!pattern.includes('?') || pattern.split('').every((ch) => ch === '?')) {
    return null;
  }
  if (pattern.startsWith('?') && /^\?[\u4e00-\u9fff]{3}$/.test(pattern)) {
    return null;
  }
  const anchors: Array<[number, string]> = [];
  for (let pos = 0; pos < pattern.length; pos++) {
    const ch = pattern[pos]!;
    if (ch !== '?') {
      anchors.push([pos, ch]);
    }
  }
  if (!anchors.length) {
    return null;
  }
  return {
    kind: QueryKind.PARTIAL_INITIAL_MASK,
    raw_q: q,
    pattern,
    width: 4,
    anchors,
  };
}

/** Port of query_grammar/rhyme.parse_rhyme_anchor_query (P1 subset) */
export function parseRhymeAnchorQuery(q: string): RhymeAnchorQuery | null {
  if (!q || q.includes(CODE_TAIL_MIDDLE) || q.includes('+') || q.includes('@') || isFramedEqualsQuery(q)) {
    return null;
  }
  if (parseDoubleWildcardRhymeQuery(q) || parseDoubleWildcardInitialQuery(q)) {
    return null;
  }

  const base = (fields: Omit<RhymeAnchorQuery, 'kind' | 'raw_q'>): RhymeAnchorQuery => ({
    kind: QueryKind.RHYME_ANCHOR,
    raw_q: q,
    ...fields,
  });

  let m = q.match(/^([\u4e00-\u9fff])=$/);
  if (m) {
    return base({
      constraint: 'final',
      anchor: m[1]!,
      anchor_pos: 0,
      slots: '',
      width: 1,
    });
  }

  m = q.match(/^=([\u4e00-\u9fff])$/);
  if (m) {
    return base({
      constraint: 'initial',
      anchor: m[1]!,
      anchor_pos: 0,
      slots: '',
      width: 1,
    });
  }

  m = q.match(/^([0-9_?%]+)([\u4e00-\u9fff])=$/);
  if (m) {
    const slots = m[1]!;
    return base({
      constraint: 'final',
      anchor: m[2]!,
      anchor_pos: slots.length,
      slots,
      width: slots.length + 1,
    });
  }

  m = q.match(/^([\u4e00-\u9fff])=([0-9_?%]+)$/);
  if (m) {
    const slots = m[2]!;
    return base({
      constraint: 'final',
      anchor: m[1]!,
      anchor_pos: 0,
      slots,
      width: slots.length + 1,
    });
  }

  m = q.match(/^=([\u4e00-\u9fff])([0-9_?%]+)$/);
  if (m) {
    const slots = m[2]!;
    return base({
      constraint: 'initial',
      anchor: m[1]!,
      anchor_pos: 0,
      slots,
      width: slots.length + 1,
    });
  }

  m = q.match(/^([0-9_?%]+)=([\u4e00-\u9fff])$/);
  if (m) {
    const slots = m[1]!;
    return base({
      constraint: 'initial',
      anchor: m[2]!,
      anchor_pos: slots.length,
      slots,
      width: slots.length + 1,
    });
  }

  return null;
}
